src/main/com/fulcrologic/fulcro/algorithms/normalized_state_helpers.cljc

```diff
@@ -0,0 +1,303 @@
+
+  The optional `cascade` parameter is a set of keywords that represent edges that should cause recursive deletes
+  (i.e. it indicates edge names that *own* something, indicating it is safe to remove those entities as well).
+
+  Returns the new state map with the entity(ies) removed."
+
+  ([state-map ident]
+   [map? eql/ident? => map?]
+   (remove-entity* state-map ident #{}))
+
+  ([state1 ident cascade]
+   [map? eql/ident? (s/coll-of keyword? :kind set?) => map?]
+   (let [tables (keep (fn [k]
+                        (let [candidate (get state1 k)]
+                          (when (and (map? candidate) (every? map? (vals candidate)))
+                            k))) (keys state1))
+
+         non-tables (keep (fn [k]
+                            (let [candidate (get state1 k)]
+                              (when (vector? candidate)
+                                [k])))
+                          (keys state1))
+
+         remove-idents-at-path (fn [state1 path]
+                                 (let [v (clojure.core/get-in state1 path)]
+                                   (cond
+                                     (int? v) state1
+                                     (= v ident) (dissoc-in state1 path)
+                                     (every? eql/ident? v) (merge/remove-ident* state1 ident path)
+                                     :else state1)))
+
+         candidate-paths (fn [state table-name]
+                           (filter (fn [a-path]
+                                     (let [vl (clojure.core/get-in state a-path)]
+                                       (if (coll? vl)
+                                         (or
+                                           (some #{ident} vl)
+                                           (= ident vl))
+                                         (= ident (take 2 a-path)))))
+                                   (normalized-paths state)))
+
+         remove-ident-from-table (fn [state1 table]
+                                   (reduce
+                                     remove-idents-at-path
+                                     state1
+                                     (concat (candidate-paths state1 (first ident)) non-tables)))
+
+         state-without-entity (->
+                                ;; remove the (non) table-nested pointers to the entity
+                                (reduce remove-ident-from-table
+                                        state1
+                                        tables)
+                                ;; remove the top-level entity
+                                (dissoc-in ident))
+
+         target-entity (get-in state1 ident)
+
+         cascaded-idents (fn [original-state target-entity cascade]
+                           (map
+                             (fn [x] (clojure.core/get-in original-state
+                                                          (conj [(first ident) ((first ident) target-entity)] x)))
```

Comment:
A `let` around this would make it a lot clearer. `(let [table-key (first ident)] ...)`